src/query/ts/m3db/encoded_step_iterator.go

```diff
@@ -0,0 +1,230 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/models"
+)
+
+type peekValue struct {
+	started  bool
+	finished bool
+	point    ts.Datapoint
+}
+
+type encodedStepIter struct {
+	mu           sync.RWMutex
+	lastBlock    bool
+	started      bool
+	currentTime  time.Time
+	err          error
+	bounds       models.Bounds
+	meta         block.Metadata
+	seriesMeta   []block.SeriesMeta
+	seriesIters  []encoding.SeriesIterator
+	seriesPeek   []peekValue
+	consolidator consolidator
+}
+
+func (b *encodedBlock) stepIter() block.StepIter {
+	cs := b.consolidation
+	consolidator := buildConsolidator(
+		time.Minute,
+		cs.bounds.StepSize,
+		cs.currentTime,
+		len(b.seriesBlockIterators),
+		cs.consolidationFn,
+	)
+	return &encodedStepIter{
+		lastBlock:    b.lastBlock,
+		currentTime:  cs.currentTime,
+		bounds:       cs.bounds,
+		meta:         b.meta,
+		seriesMeta:   b.seriesMetas,
+		seriesIters:  b.seriesBlockIterators,
+		consolidator: consolidator,
+	}
+}
+
+type encodedStep struct {
+	time   time.Time
+	values []float64
+}
+
+func (s *encodedStep) Time() time.Time   { return s.time }
+func (s *encodedStep) Values() []float64 { return s.values }
+
+func (it *encodedStepIter) Current() (block.Step, error) {
+	it.mu.RLock()
+	currentTime := it.currentTime
+	if !it.started || currentTime.After(it.bounds.End()) {
+		it.mu.RUnlock()
+		return nil, fmt.Errorf("out of bounds")
+	}
+
+	if it.err != nil {
+		it.mu.RUnlock()
+		return nil, it.err
+	}
+
+	step := &encodedStep{
+		time:   currentTime,
+		values: it.consolidator.consolidate(),
+	}
+
+	it.mu.RUnlock()
+	return step, nil
+}
+
+func (it *encodedStepIter) nextConsolidatedForStep(i int) {
+	peek := it.seriesPeek[i]
+	if peek.finished {
+		// No next value in this iterator
+		return
+	}
+
+	if peek.started {
```

Comment:
I think it's just a rather complex set of logic; not sure if it would make it much clearer, and we may just be kicking complexity a bit further down the stack

Reply:
That's fair, though I would argue peek in isolation is actually pretty simple, whereas combined with everything else it's harder to reason about. That said, the function's still under 100 lines or so, so I can live with the complexity here.